src/Generating/FinishGen.cpp

```diff
@@ -776,36 +837,26 @@ void cFinishGenSprinkleFoliage::GenFinish(cChunkDesc & a_ChunkDesc)
 					{
 						a_ChunkDesc.SetBlockTypeMeta(x, ++Top, z, E_BLOCK_TALL_GRASS, E_META_TALL_GRASS_GRASS);
 					}
-					else if (TryAddSugarcane(a_ChunkDesc, x, Top, z))
-					{
-						++Top;
-					}
 					else if ((val1 > 0.5) && (val2 < -0.5))
 					{
 						a_ChunkDesc.SetBlockTypeMeta(x, ++Top, z, E_BLOCK_PUMPKIN, static_cast<int>(val3 * 8) % 4);
 					}
+					else
+					{
+						TryAddSugarcane(a_ChunkDesc, x, Top, z);
+					}
 					break;
 				}  // case E_BLOCK_GRASS
 
 				case E_BLOCK_SAND:
 				{
-					int y = Top + 1;
-					if (
-						(x > 0) && (x < cChunkDef::Width - 1) &&
-						(z > 0) && (z < cChunkDef::Width - 1) &&
-						(val1 + val2 > 0.5f) &&
-						(a_ChunkDesc.GetBlockType(x + 1, y, z)     == E_BLOCK_AIR) &&
-						(a_ChunkDesc.GetBlockType(x - 1, y, z)     == E_BLOCK_AIR) &&
-						(a_ChunkDesc.GetBlockType(x,     y, z + 1) == E_BLOCK_AIR) &&
-						(a_ChunkDesc.GetBlockType(x,     y, z - 1) == E_BLOCK_AIR) &&
-						IsDesertVariant(a_ChunkDesc.GetBiome(x, z))
-					)
+					if (val1 + val2 > 0.5f && IsDesertVariant(a_ChunkDesc.GetBiome(x, z)))
 					{
-						a_ChunkDesc.SetBlockType(x, ++Top, z, E_BLOCK_CACTUS);
+						TryAddCactus(a_ChunkDesc, x, Top, z);
```

Comment:
A seperate `if` condition is better, because you need the same for the sugarcane. E.g. if sugarcane can't be placed, try cacti.
The best construct I currently can think of:
```cpp
if(val1 + val2 > 0.5f)
{
  if(!TryAddCactus(a_ChunkDesc, x, Top, z))
  {
    TryAddSugarcane(a_ChunkDesc, x, Top, z);
  }
}
else
{
  if(!TryAddSugarcane(a_ChunkDesc, x, Top, z))
  {
    TryAddCactus(a_ChunkDesc, x, Top, z);
  }
}
```
Note: The code above implies, that `IsDesertVariant` is checked inside the `TryAddCactus` function.

Reply:
> if sugarcane can't be placed, try cacti

It didn't behave like that before so it shouldn't be changed to do that.  This PR should only change the height of sugarcanes and cacti but not where they generate.

Reply:
Yep, you're right... I overlooked this behaviour in the old code.
But please don't mix the `TryAddX` functions into the `val1 + val2` condition, when fixing this:
```cpp
if(val1 + val2 <= 0.5f || !TryAddCactus(a_ChunkDesc, x, Top, z))
{
  TryAddSugarcane(a_ChunkDesc, x, Top, z);
}
```